### PR TITLE
Analyze clickbench performance with explain plans and pg_stat_statements

### PR DESCRIPTION
**Pull request by Bodobolero**
## Problem

To understand differences in performance between neon, aurora and rds we want to collect explain analyze plans and pg_stat_statements for selected benchmarking runs

## Summary of changes

Add workflow input options to collect explain and pg_stat_statements for benchmarking workflow 

## Checklist before requesting a review

- [x ] I have performed a self-review of my code.
- [N/A ] If it is a core feature, I have added thorough tests. - it is not a core feature
- [N/A] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [N/A] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: .github/workflows/benchmarking.yml [hunk 1]
```diff
@@ -120,66 +135,66 @@ jobs:
...
+          if [ "$(date +%A)" = "Saturday" ] || ${{ github.event.inputs.run_AWS_RDS_AND_AURORA }}; then
+            matrix=$(echo "$matrix" | jq '.include += [{ "platform": "rds-postgres" },
+                                                      { "platform": "rds-aurora"   }]')
+          fi
+
+          echo "matrix=$(echo "$matrix" | jq --compact-output '.')" >> $GITHUB_OUTPUT
+
+      - name: Generate matrix for TPC-H benchmarks
+        id: tpch-compare-matrix
+        run: |
```
[the remark this answers is not in the document]
Comment (Bodobolero) [reply .github/workflows/benchmarking.yml:179]: The variables used in the run section are defined as type "boolean" and thus can only have values true or false and a shell injection is thus not possible:

In GitHub Actions, when triggering a workflow via the workflow_dispatch event, the values users can pass for inputs are constrained by the input type defined in the workflow file. For inputs of specific types like boolean, choice, or environment, GitHub enforces the expected value format:

Via GitHub UI: When a user triggers a workflow through the GitHub web interface, the UI provides controls (like text boxes, dropdowns, or toggles) based on the input type. For a boolean, it typically provides a checkbox or a toggle. This UI control restricts the user to the valid options (i.e., true or false for booleans).
Using GitHub API: When triggering a workflow through the GitHub API, the data passed in the API request must match the expected types defined in the workflow file. If the input type is boolean, the API expects a boolean value (true or false). GitHub's validation should reject values that don't match the expected type.

File: .github/workflows/benchmarking.yml [hunk 2]
```diff
@@ -115,208 +130,208 @@ jobs:
...
+      #     if [ "$(date +%A)" = "Saturday" ] || ${{ github.event.inputs.run_AWS_RDS_AND_AURORA }}; then
+      #       matrix=$(echo "$matrix" | jq '.include += [{ "platform": "rds-postgres", "db_size": "10gb"},
+      #                                                { "platform": "rds-aurora",   "db_size": "50gb"}]')
+      #     fi
+
+      #     echo "matrix=$(echo "$matrix" | jq --compact-output '.')" >> $GITHUB_OUTPUT
+
+      - name: Generate matrix for OLAP benchmarks
+        id: olap-compare-matrix
+        run: |
```
[the remark this answers is not in the document]
Comment (Bodobolero) [reply .github/workflows/benchmarking.yml:163]: The variables used in the run section are defined as type "boolean" and thus can only have values true or false and a shell injection is thus not possible:

In GitHub Actions, when triggering a workflow via the workflow_dispatch event, the values users can pass for inputs are constrained by the input type defined in the workflow file. For inputs of specific types like boolean, choice, or environment, GitHub enforces the expected value format:

Via GitHub UI: When a user triggers a workflow through the GitHub web interface, the UI provides controls (like text boxes, dropdowns, or toggles) based on the input type. For a boolean, it typically provides a checkbox or a toggle. This UI control restricts the user to the valid options (i.e., true or false for booleans).
Using GitHub API: When triggering a workflow through the GitHub API, the data passed in the API request must match the expected types defined in the workflow file. If the input type is boolean, the API expects a boolean value (true or false). GitHub's validation should reject values that don't match the expected type.

File: .github/workflows/benchmarking.yml [hunk 3]
```diff
@@ -120,63 +135,63 @@ jobs:
...
+          matrix='{
+            "platform": [
+              "neon-captest-reuse"
+            ]
+          }'
+
+          if [ "$(date +%A)" = "Saturday" ] || ${{ github.event.inputs.run_AWS_RDS_AND_AURORA }}; then
```
Comment (bayandin) [inline .github/workflows/benchmarking.yml:170]: For scheduled workflow runs template `${{ github.event.inputs.run_AWS_RDS_AND_AURORA }}` is evaluated as a blank space, and this makes a bash script invalid. 
Inputs variables are evaluated only for runs that are triggered by `workflow_dispatch` event (and their default values as well, i.e. for scheduled runs you can't rely on it):

```bash
$ if [ "$(date +%A)" = "Saturday" ] || ; then echo true; else echo false; fi
bash: syntax error near unexpected token `;'
```

I would recommend use an env var to workaround this, something like:
```yaml
env:
  RUN_AWS_RDS_AND_AURORA: ${{ github.event.inputs.run_AWS_RDS_AND_AURORA || 'false' }}
```

And then use `${RUN_AWS_RDS_AND_AURORA}` in the bash part

File: .github/workflows/benchmarking.yml [hunk 4]
```diff
@@ -120,63 +135,63 @@ jobs:
...
-          ]
-        }'
-
-        if [ "$(date +%A)" = "Saturday" ]; then
-          matrix=$(echo "$matrix" | jq '.include += [{ "platform": "rds-postgres", "scale": "10" },
-                                                   { "platform": "rds-aurora",   "scale": "10" }]')
-        fi
-
-        echo "matrix=$(echo "$matrix" | jq --compact-output '.')" >> $GITHUB_OUTPUT
+      - name: Generate matrix for pgbench benchmark
```
Comment (bayandin) [inline .github/workflows/benchmarking.yml:138]: I see some noop style changes in the yaml file, let's move it to a separate PR

File: test_runner/performance/test_perf_olap.py [hunk 5]
```diff
@@ -16,6 +16,21 @@ class LabelledQuery:
...
+# Theoretically this could be in a module or session scope fixture,
+# however the code depends on other fixtures that have function scope
+@pytest.mark.remote_cluster
+def test_clickbench_create_pg_stat_statements(remote_compare: RemoteCompare):
+    if os.getenv('TEST_OLAP_COLLECT_PG_STAT_STATEMENTS', 'false').lower() == 'true':
+        log.info("Creating extension pg_stat_statements")
+        query =  LabelledQuery("Q_CREATE_EXTENSION", r"CREATE EXTENSION pg_stat_statements;")
+        run_psql_once_without_explain(remote_compare, query)
+        log.info("Reset pg_stat_statements")
+        query =  LabelledQuery("Q_RESET", r"SELECT pg_stat_statements_reset();")
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:30]: We don't execute this query

File: test_runner/performance/test_perf_olap.py [hunk 6]
```diff
@@ -16,6 +16,21 @@ class LabelledQuery:
     label: str
     query: str
 
+# create extension pg_stat_statements before all tests in this module if it does not exist
+# and TEST_OLAP_COLLECT_PG_STAT_STATEMENTS is set to true (default false)
+# Theoretically this could be in a module or session scope fixture,
+# however the code depends on other fixtures that have function scope
+@pytest.mark.remote_cluster
+def test_clickbench_create_pg_stat_statements(remote_compare: RemoteCompare):
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:24]: I would add that we rely on this test to be executed first to a comment

File: .github/workflows/benchmarking.yml [hunk 7]
```diff
@@ -337,83 +352,87 @@ jobs:
       POSTGRES_DISTRIB_DIR: /tmp/neon/pg_install
       DEFAULT_PG_VERSION: 14
       TEST_OUTPUT: /tmp/test_output
+      TEST_OLAP_COLLECT_EXPLAIN: ${{ github.event.inputs.collect_olap_explain }}
+      TEST_OLAP_COLLECT_PG_STAT_STATEMENTS: ${{ github.event.inputs.collect_pg_stat_statements }}
```
Comment (bayandin) [inline .github/workflows/benchmarking.yml:356]: JFYI, this will be evaluated to empty strings for scheduled runs. If you want to default them to `false`, then you need to do something like this

```suggestion
      TEST_OLAP_COLLECT_EXPLAIN: ${{ github.event.inputs.collect_olap_explain || 'false' }}
      TEST_OLAP_COLLECT_PG_STAT_STATEMENTS: ${{ github.event.inputs.collect_pg_stat_statements || 'false' }}
```

File: test_runner/performance/test_perf_olap.py [hunk 8]
```diff
@@ -108,7 +125,28 @@ def run_psql(env: RemoteCompare, labelled_query: LabelledQuery, times: int) -> N
         log.info(f"Run {run}/{times}")
         with env.zenbenchmark.record_duration(f"{label}/{run}"):
             env.pg_bin.run_capture(["psql", connstr, "-c", query], env=environ)
+    if os.getenv('TEST_OLAP_COLLECT_EXPLAIN', 'false').lower() == 'true':
+        log.info(f"Explaining query {label}")
+        run += 1
+        with env.zenbenchmark.record_duration(f"{label}/EXPLAIN"):
+            env.pg_bin.run_capture(["psql", connstr, "-c", EXPLAIN_STRING+query], env=environ)
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:132]:
```suggestion
            env.pg_bin.run_capture(["psql", connstr, "-c", f"{EXPLAIN_STRING} {query}"], env=environ)
```

And then we can delete an extra space from the end of `EXPLAIN_STRING`

File: test_runner/performance/test_perf_olap.py [hunk 9]
```diff
@@ -108,7 +125,28 @@ def run_psql(env: RemoteCompare, labelled_query: LabelledQuery, times: int) -> N
         log.info(f"Run {run}/{times}")
         with env.zenbenchmark.record_duration(f"{label}/{run}"):
             env.pg_bin.run_capture(["psql", connstr, "-c", query], env=environ)
+    if os.getenv('TEST_OLAP_COLLECT_EXPLAIN', 'false').lower() == 'true':
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:128]: I prefer to add an argument to the method (something like `explain: bool = False`), use it in the body, and have `os.getenv('TEST_OLAP_COLLECT_EXPLAIN', 'false').lower() == 'true'` part in a test itself.
This should eliminate the need of an additional `run_psql_once_without_explain` method

**Review by bayandin — COMMENTED** [review]
Great idea!
Just a couple comments regarding the workflow and Python nitpicks

File: .github/workflows/benchmarking.yml (hunk 3 above)
Comment (Bodobolero) [reply .github/workflows/benchmarking.yml:170]: Good catch.

File: .github/workflows/benchmarking.yml (hunk 4 above)
Comment (Bodobolero) [reply .github/workflows/benchmarking.yml:138]: Sorry for that. My visual studio code extension for Yaml was set to auto format on save by mistake. My diff editor ignored whitespace so I didn't even notice some unintended changes.

File: .github/workflows/benchmarking.yml (hunk 7 above)
Comment (Bodobolero) [reply .github/workflows/benchmarking.yml:356]: good catch

File: test_runner/performance/test_perf_olap.py (hunk 6 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py:24]: I had "before all tests" but I will emphasize it more

File: test_runner/performance/test_perf_olap.py (hunk 5 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py:30]: oops!

File: test_runner/performance/test_perf_olap.py (hunk 9 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py:128]: Yes your suggestion is much better style (DRY)

File: test_runner/performance/test_perf_olap.py (hunk 8 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py:132]: ok

File: test_runner/performance/test_perf_olap.py [hunk 10]
```diff
@@ -17,6 +17,24 @@ class LabelledQuery:
     query: str
 
 
+# This must run before all tests in this module
+# create extension pg_stat_statements if it does not exist
+# and TEST_OLAP_COLLECT_PG_STAT_STATEMENTS is set to true (default false)
+# Theoretically this could be in a module or session scope fixture,
+# however the code depends on other fixtures that have function scope
+@pytest.mark.remote_cluster
+def test_clickbench_create_pg_stat_statements(remote_compare: RemoteCompare):
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:26]: There's one more thing that we can use here is `skipif`, then we don't need to have a condition in the function itself. 

```suggestion
@pytest.mark.skipif(os.getenv('TEST_OLAP_COLLECT_PG_STAT_STATEMENTS', 'false').lower() == 'true', reason="<put the reason here>")
def test_clickbench_create_pg_stat_statements(remote_compare: RemoteCompare):
```

And the same about `test_clickbench_collect_pg_stat_statements`

File: test_runner/performance/test_perf_olap.py [hunk 11]
```diff
@@ -88,7 +108,10 @@ def get_scale() -> List[str]:
     return [scale]
 
 
-def run_psql(env: RemoteCompare, labelled_query: LabelledQuery, times: int) -> None:
+# run the query times times plus once with EXPLAIN VERBOSE if TEST_OLAP_COLLECT_EXPLAIN is set to true (default false)
```
Comment (bayandin) [inline test_runner/performance/test_perf_olap.py:111]: This comment might be a bit outdated, we don't really check `TEST_OLAP_COLLECT_EXPLAIN` here anymore

File: test_runner/performance/test_perf_olap.py (hunk 10 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py:27]: OK, however afaik the logic is reversed:
```
@pytest.mark.skipif(os.getenv('TEST_OLAP_COLLECT_PG_STAT_STATEMENTS', 'false').lower() == 'false', reason="Skipping - Creating extension pg_stat_statements")
```

File: test_runner/performance/test_perf_olap.py (hunk 11 above)
Comment (Bodobolero) [reply test_runner/performance/test_perf_olap.py]: fixed

**Review by bayandin — APPROVED** [review]

**Review by bayandin — APPROVED** [review]
Awesome!

Please use "Squash and merge" to merge the PR (i.e. not "Rebase and merge")
